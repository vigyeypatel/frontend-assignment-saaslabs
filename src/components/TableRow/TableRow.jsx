import React from "react";
import formatColumnName from "../../utils/index"

/**
 * A TableRow component that dynamically renders each row's data based on columns.
 * 
 * @param {Object} item - The data for the current row.
 * @param {Array} columns - The column names to display.
 */
const TableRow = ({ item, columns }) => {
  return (
    <tr>
      {columns.map((column) => {
        // Access data dynamically using column name
        const colName = formatColumnName(column);
        const cellData = item[colName];
        return <td key={column.toString()}>{cellData}</td>;
      })}
    </tr>
  );
};

export default TableRow;