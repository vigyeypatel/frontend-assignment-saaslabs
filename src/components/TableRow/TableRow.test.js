import { render, screen } from "@testing-library/react";
import TableRow from "./TableRow";
import formatColumnName from "../../utils/index";

jest.mock("../../utils/index", () => jest.fn((column) => column.toLowerCase()));

describe("TableRow Component", () => {
  const columns = ["Date", "Amount", "Category", "Description"];
  const item = {
    date: "2023-12-25",
    amount: "$100",
    category: "Groceries",
    description: "Christmas shopping",
  };

  test("renders correct number of cells", () => {
    render(<TableRow item={item} columns={columns} />);
    const cells = screen.getAllByRole("cell");
    expect(cells).toHaveLength(columns.length);
  });

  test("displays correct data in each cell", () => {
    render(<TableRow item={item} columns={columns} />);
    expect(screen.getByText("2023-12-25")).toBeInTheDocument();
    expect(screen.getByText("$100")).toBeInTheDocument();
    expect(screen.getByText("Groceries")).toBeInTheDocument();
    expect(screen.getByText("Christmas shopping")).toBeInTheDocument();
  });

  test("calls formatColumnName for each column", () => {
    render(<TableRow item={item} columns={columns} />);
    expect(formatColumnName).toHaveBeenCalledTimes(columns.length);
    columns.forEach((col) => {
      expect(formatColumnName).toHaveBeenCalledWith(col);
    });
  });

  test("handles missing columns or item gracefully", () => {
    const { container } = render(<TableRow item={{}} columns={[]} />);
    expect(container.querySelector("tr")).toBeInTheDocument();
    expect(container.querySelectorAll("td")).toHaveLength(0);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<TableRow item={item} columns={columns} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
