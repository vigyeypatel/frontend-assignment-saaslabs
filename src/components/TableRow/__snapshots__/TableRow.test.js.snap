// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow Component matches snapshot 1`] = `<DocumentFragment />`;
